provider/provider.go

```diff
@@ -0,0 +1,270 @@
+
+	// PlanResourceChange takes the current state and proposed state of a
+	// resource, and returns the planned final state.
+	PlanResourceChange(PlanResourceChangeRequest) PlanResourceChangeResponse
+
+	// ApplyResourceChange takes the planned state for a resource, which may
+	// yet contain unknown computed values, and applies the changes returning
+	// the final state.
+	ApplyResourceChange(ApplyResourceChangeRequest) ApplyResourceChangeResponse
+
+	// ImportResourceState requests that the given resource be imported.
+	ImportResourceState(ImportResourceStateRequest) ImportResourceStateResponse
+
+	// ReadDataSource returns the data source's current state.
+	ReadDataSource(ReadDataSourceRequest) ReadDataSourceResponse
+}
+
+type GetSchemaResponse struct {
+	// Provider is the schema for the provider itself.
+	Provider *configschema.Block
+
+	// ResourceTypes map the resource type name to that type's schema.
+	ResourceTypes map[string]*configschema.Block
+
+	// DataSources maps the data source name to that data source's schema.
+	DataSources map[string]*configschema.Block
+
+	// Diagnostics contains any warnings or errors from the method call.
+	Diagnostics tfdiags.Diagnostics
+}
+
+type ValidateProviderConfigRequest struct {
+	// Config is the complete configuration value for the provider.
+	Config cty.Value
+}
+
+type ValidateProviderConfigResponse struct {
+	// Diagnostics contains any warnings or errors from the method call.
+	Diagnostics tfdiags.Diagnostics
+}
+
+type ValidateResourceTypeConfigRequest struct {
+	// TypeName is the name of the resource type to validate.
+	TypeName string
+
+	// Config is the configuration value to validate, which may contain unknown
+	// values.
+	Config cty.Value
+}
+
+type ValidateResourceTypeConfigResponse struct {
+	// Diagnostics contains any warnings or errors from the method call.
+	Diagnostics tfdiags.Diagnostics
+}
+
+type ValidateDataSourceConfigRequest struct {
+	// TypeName is the name of the data source type to validate.
+	TypeName string
+
+	// Config is the configuration value to validate, which may contain unknown
+	// values.
+	Config cty.Value
+}
+
+type ValidateDataSourceConfigResponse struct {
+	// Diagnostics contains any warnings or errors from the method call.
+	Diagnostics tfdiags.Diagnostics
+}
+
+type UpgradeResourceStateRequest struct {
+	// TypeName is the name of the resource type being upgraded
+	TypeName string
+
+	// Version is version of the schema that created the current state.
+	Version int
+
+	// PriorStateRaw is the state that needs to be upgraded to match the current
+	// schema version. Because the schema is unknown, this contains only the
+	// raw bytes as stored in the state. It is up to the provider to interpret
+	// the payload and return a state encoded with the current schema.
+	PriorStateRaw []byte
+}
+
+type UpgradeResourceStateResponse struct {
+	// State is the newly upgraded resource state.
+	State cty.Value
```

Comment:
In all of the other cases we've used an adjective with state to represent how it changes as it moves through this workflow, which makes me tempted to name this `UpgradedState` for consistency.